importers/rules.go

```diff
@@ -0,0 +1,256 @@
+type RuleImporter struct {
+	logger log.Logger
+
+	config RuleConfig
+	groups map[string]*rules.Group
+
+	apiClient v1.API
+
+	writer *blocks.MultiWriter
+}
+
+// RuleConfig is the config for the rule importer
+type RuleConfig struct {
+	Start        string
+	End          string
+	EvalInterval time.Duration
+	URL          string
+}
+
+// NewRuleImporter creates a new rule importer
+func NewRuleImporter(logger log.Logger, config RuleConfig) *RuleImporter {
+	return &RuleImporter{
+		config: config,
+	}
+}
+
+// Init initializes the rule importer which creates a new block writer
+// and creates an Prometheus API client
+func (importer *RuleImporter) Init() error {
+	// create new block writer
+	newBlockDir, err := ioutil.TempDir("", "rule_blocks")
+	if err != nil {
+		return err
+	}
+	importer.writer = blocks.NewMultiWriter(importer.logger, newBlockDir, importer.config.EvalInterval.Nanoseconds())
+
+	// create api client
+	config := api.Config{
+		Address: importer.config.URL,
+	}
+	c, err := api.NewClient(config)
+	if err != nil {
+		return err
+	}
+	importer.apiClient = v1.NewAPI(c)
+	return nil
+}
+
+// Close cleans up any open resources
+func (importer *RuleImporter) Close() error {
+	return importer.writer.Close()
+}
+
+// Parse parses the groups and rules from a list of rules files
+func (importer *RuleImporter) Parse(ctx context.Context, files []string) (errs []error) {
+	groups := make(map[string]*rules.Group)
+
+	for _, file := range files {
+		ruleGroups, errs := rulefmt.ParseFile(file)
+		if errs != nil {
+			return errs
+		}
+
+		for _, ruleGroup := range ruleGroups.Groups {
+			itv := importer.config.EvalInterval
+			if ruleGroup.Interval != 0 {
+				itv = time.Duration(ruleGroup.Interval)
+			}
+
+			rulez := make([]rules.Rule, 0, len(ruleGroup.Rules))
+			for _, r := range ruleGroup.Rules {
+				expr, err := parser.ParseExpr(r.Expr.Value)
+				if err != nil {
+					return []error{errors.Wrap(err, file)}
+				}
+
+				rulez = append(rulez, rules.NewRecordingRule(
+					r.Record.Value,
+					expr,
+					labels.FromMap(r.Labels),
+				))
+			}
+
+			groups[file+";"+ruleGroup.Name] = rules.NewGroup(rules.GroupOptions{
+				Name:     ruleGroup.Name,
+				File:     file,
+				Interval: itv,
+				Rules:    rulez,
+			})
+		}
+	}
+
+	importer.groups = groups
+	return errs
+}
+
+// ImportAll evaluates all the groups and rules and creates new time series
+// and stores in new blocks
+func (importer *RuleImporter) ImportAll(ctx context.Context) []error {
+	var errs = []error{}
+	for _, group := range importer.groups {
+		for _, rule := range group.Rules() {
+			err := importer.ImportRule(ctx, rule)
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+	err := importer.CreateBlocks()
+	if err != nil {
+		errs = append(errs, err)
+	}
+	return errs
+}
+
+func (importer *RuleImporter) queryFn(ctx context.Context, q string, t time.Time) (promql.Vector, error) {
+	val, warnings, err := importer.apiClient.Query(ctx, q, t)
```

Comment:
Why is this using query rather than query range?

Reply:
Query range does not stream the response.

Reply:
We don't need streamed responses, at most query range will return 50M samples. We want to leave the query evaluation up to the other end, not do it inside promtool.

Reply:
> We don't need streamed responses, at most query range will return 50M samples.

What do you mean? This is worse than lack of streaming as this means we cannot even do anything beyond 50M samples. Imagine someone backfilling 1y for single series. Although such a thing we might be happy to just reject.

Otherwise query range means less requests to the server indeed. We can definitely think about using the query range then, however, there is a nontrivial issue with `step` adjusting for query range to match the actual eval interval we want. :thinking: 

As per if I query `[0 - 100]` with step 2 will it start at 0 always?

Reply:
I'm not proposing to pull in raw series and do PromQL evaluation in promtool - that'd be very inefficient. Rather get the Prometheus on the other end to do all the heavy lifting.

Reply:
I'm will give queryRange API and see how that works out. Pushing a commit with that change shortly